src/gui/activitywidget.cpp

```diff
@@ -400,6 +214,154 @@ void ActivityWidget::slotOpenFile(QModelIndex indx)
     }
 }
 
+// GUI: Display the notifications
+void ActivityWidget::slotBuildNotificationDisplay(const ActivityList& list)
+{
+    int newGuiLogs = 0;
+
+    QHash<QString, int> accNotified;
+
+    foreach( auto activity, list ) {
+        NotificationWidget *widget = 0;
+
+        if( _widgetForNotifId.contains(activity._id) ) {
+            widget = _widgetForNotifId[activity._id];
+        } else {
+            widget = new NotificationWidget(this);
+            connect(widget, SIGNAL(sendNotificationRequest(QString, QString, QString)),
+                    this, SLOT(slotSendNotificationRequest(QString, QString, QString)));
+            _notificationsLayout->addWidget(widget);
+            // _ui->_notifyScroll->setMinimumHeight( widget->height());
+            _ui->_notifyScroll->setSizeAdjustPolicy(QAbstractScrollArea::AdjustToContentsOnFirstShow);
+            _widgetForNotifId[activity._id] = widget;
+        }
+
+        widget->setAccountName( activity._accName );
+        widget->setActivity( activity );
+
+        // handle gui logs. In order to NOT annoy the user with every fetching of the
+        // notifications the notification id is stored in a Set. Only if an id
+        // is not in the set, it qualifies for guiLog.
+        // Important: The _guiLoggedNotifications set must be wiped regularly which
+        // will repeat the gui log.
+
+        // after one hour, clear the gui log notification store
+        if( _guiLogTimer.elapsed() > 60*60*1000 ) {
+            _guiLoggedNotifications.clear();
+        }
+        if( !_guiLoggedNotifications.contains(activity._id)) {
+            newGuiLogs++;
+            QString host = activity._accName;
+            // store the name of the account that sends the notification to be
+            // able to add it to the tray notification
+            // remove the user name from the account as that is not accurate here.
+            int indx = host.indexOf(QChar('@'));
+            if( indx>-1 ) {
+                host.remove(0, 1+indx);
+            }
+            if( !host.isEmpty() ) {
+                if( accNotified.contains(host)) {
+                    accNotified[host] = accNotified[host]+1;
+                } else {
+                    accNotified[host] = 1;
+                }
+            }
+            _guiLoggedNotifications.insert(activity._id);
+        }
+    }
+
+    _ui->_notifyLabel->setHidden( _widgetForNotifId.isEmpty() );
+    _ui->_notifyScroll->setHidden( _widgetForNotifId.isEmpty() );
+
+    if( newGuiLogs > 0 ) {
+        // restart the gui log timer now that we show a notification
+        _guiLogTimer.restart();
+
+        // Assemble a tray notification
```

Comment:
I don't think you can build strings this way. Even with translator comments This will break in all kind of funny ways e.g. with RTL languages. Also, with too many notifiers, the message cannot be parsed fast enough, so we should probably limit the amount of displayed notification sources (?) as well. You at have to account for three cases, with the n=2 one being optional:

* `n =1` (You received %n notifications from %1).arg(notifier)
* `n = 2` (You received %n notifications from %1 and %2).arg(notifiers).
* `n > 2` (You received %n notifications from %1, %2 and others).arg(notifiers.at(0), notifiers.at(1). (ideally, we would specify a quantity for "others", but we cannot have more than one plural form in one tr()...